### ISSUE TITLE
How to best report translation failures

### ISSUE THREAD

**Issue by jpalmerpivotal** [imsweb/validation-translation#4]
This is an open ended issue about the best way for me to report translation failures. There were a decent number for some states, and I was wondering what would be the best way to report them for your convenience, either as a single issue per type of error (there were a number of regexs), a mega issue or what.
We would also be happy to try and PR some fixes if that would suit you.

Thanks, James.

**Comment by jpalmerpivotal** [comment]
Just going to start compiling example failures in this comment per file:

NAACCR v22A

> !!! Unable to parse 'Autopsy Only, RX, Schema ID, Primary Site, 2018 (NPCR)': class com.imsweb.validation.translation.language.entity.IfThenElseStatement cannot be cast to class com.imsweb.validation.translation.language.entity.ParsedTree (com.imsweb.validation.translation.language.entity.IfThenElseStatement and com.imsweb.validation.translation.language.entity.ParsedTree are in unnamed module of loader 'app')

Illinois v21B

  > !!! Unable to parse 'IL-CenRgEdit Warning!! Appendix Carcinoid, Diagnosis Date ': class com.imsweb.validation.translation.language.entity.OperationExpression cannot be cast to class com.imsweb.validation.translation.language.entity.ParsedTree (com.imsweb.validation.translation.language.entity.OperationExpression and com.imsweb.validation.translation.language.entity.ParsedTree are in unnamed module of loader 'app')

New Jersey v21B was perfect

PA 21B (pennsylvania was particularly rough)
  > !!! Unable to parse 'Z_County at DX, ADDR at DX-State': Unable to find field 'County at DX' // seems like a typo in the metafile, and these 2 edits are not in any sets
  > !!! Unable to parse 'Z_County at DX, State--MCD Code': Unable to find field 'County at DX'
  > !!! Unable to parse 'Z_Date of 1st Contact': class com.imsweb.validation.translation.language.entity.IfThenElseStatement cannot be cast to class com.imsweb.validation.translation.language.entity.ParsedTree (com.imsweb.validation.translation.language.entity.IfThenElseStatement and com.imsweb.validation.translation.language.entity.ParsedTree are in unnamed module of loader 'app')
  > !!! Unable to parse 'Z_Date of Inpt Disch, Date of 1st Contact': Unable to translate regex '        '
  > !!! Unable to parse 'Z_Date of Last Contact': class com.imsweb.validation.translation.language.entity.IfThenElseStatement cannot be cast to class com.imsweb.validation.translation.language.entity.ParsedTree (com.imsweb.validation.translation.language.entity.IfThenElseStatement and com.imsweb.validation.translation.language.entity.ParsedTree are in unnamed module of loader 'app')
  > !!! Unable to parse 'Z_Histology ICDO3-Type & Behavior': Unable to find message #501 on Z_Histology ICDO3-Type & Behavior
  > !!! Unable to parse 'Z_Institution Referred From': Unable to translate regex '          '
  > !!! Unable to parse 'Z_PA3_CurRepHosp': Unable to translate regex '          '
  > !!! Unable to parse 'Z_Physician--Follow-Up': Unable to translate regex '        '
  > !!! Unable to parse 'Z_Primary Site, Behavior Code ICDO2': Unable to translate regex ' '
  > !!! Unable to parse 'Z_Primary Site, Behavior Code ICDO3': Unable to translate regex ' '
  > !!! Unable to parse 'Z_Reg Nodes, Date of DX': Unable to translate regex '        '
  > !!! Unable to parse 'Z_Rep Fac, County of Death, DCStateFile': Unable to translate regex '      '
  > !!! Unable to parse 'Z_RX Date--Systemic, Date Flag  (2003-2009)': Unable to translate regex '        '
  > !!! Unable to parse 'Z_Summary Stage 2000, Date of Diagnosis': Unable to translate regex ' '
  > !!! Unable to parse 'Z_Summary Stage 77, Date of Diagnosis': Unable to translate regex ' '
  > !!! Unable to parse 'Z_Type Rep Source, County of Death, DCStateFileNum': Unable to translate regex '     '
  > !!! Unable to parse 'Z_Vendor Name': class com.imsweb.validation.translation.language.entity.IfThenElseStatement cannot be cast to class com.imsweb.validation.translation.language.entity.ParsedTree (com.imsweb.validation.translation.language.entity.IfThenElseStatement and com.imsweb.validation.translation.language.entity.ParsedTree are in unnamed module of loader 'app')
  > !!! Unable to parse 'ZZ_CS Items - PCR Required - SSF 1': Unable to find message #3374 on ZZ_CS Items - PCR Required - SSF 1
  > !!! Unable to parse 'ZZ_CS Items - PCR Required - SSF 10': Unable to find message #3374 on ZZ_CS Items - PCR Required - SSF 10
  > !!! Unable to parse 'ZZ_CS Items - PCR Required - SSF 11': Unable to find message #3374 on ZZ_CS Items - PCR Required - SSF 11
  > !!! Unable to parse 'ZZ_CS Items - PCR Required - SSF 12': Unable to find message #3374 on ZZ_CS Items - PCR Required - SSF 12
  > !!! Unable to parse 'ZZ_CS Items - PCR Required - SSF 13': Unable to find message #3374 on ZZ_CS Items - PCR Required - SSF 13
  > !!! Unable to parse 'ZZ_CS Items - PCR Required - SSF 14': Unable to find message #3374 on ZZ_CS Items - PCR Required - SSF 14
  > !!! Unable to parse 'ZZ_CS Items - PCR Required - SSF 16': Unable to find message #3374 on ZZ_CS Items - PCR Required - SSF 16
  > !!! Unable to parse 'ZZ_CS Items - PCR Required - SSF 17': Unable to find message #3374 on ZZ_CS Items - PCR Required - SSF 17
  > !!! Unable to parse 'ZZ_CS Items - PCR Required - SSF 2': Unable to find message #3374 on ZZ_CS Items - PCR Required - SSF 2
  > !!! Unable to parse 'ZZ_CS Items - PCR Required - SSF 3': Unable to find message #3374 on ZZ_CS Items - PCR Required - SSF 3
  > !!! Unable to parse 'ZZ_CS Items - PCR Required - SSF 4': Unable to find message #3375 on ZZ_CS Items - PCR Required - SSF 4
  > !!! Unable to parse 'ZZ_CS Items - PCR Required - SSF 5': Unable to find message #3374 on ZZ_CS Items - PCR Required - SSF 5
  > !!! Unable to parse 'ZZ_CS Items - PCR Required - SSF 6': Unable to find message #3374 on ZZ_CS Items - PCR Required - SSF 6
  > !!! Unable to parse 'ZZ_CS Items - PCR Required - SSF 7': Unable to find message #3374 on ZZ_CS Items - PCR Required - SSF 7
  > !!! Unable to parse 'ZZ_CS Items - PCR Required - SSF 8': Unable to find message #3374 on ZZ_CS Items - PCR Required - SSF 8
  > !!! Unable to parse 'ZZ_CS Items - PCR Required - SSF 9': Unable to find message #3374 on ZZ_CS Items - PCR Required - SSF 9
  > !!! Unable to parse 'ZZ_CS Items,Date of DX,Date of 1st Contact': Unable to translate regex '        '
  > !!! Unable to parse 'ZZ_Date Case Completed-CoC': Unable to find field 'Date Case Completed-CoC'
  
  Wisconsin was blocked on the `INTEGER` issue.

**Comment by depryf** [comment]
I think one issue per "problem" would work better since the fix will usually be very different (and some might not be real translation problems  or might not be fixable). 

I was away for a week and I have many other priorities competing with this. So I will review all this when I can.

One thing you probably already noticed is that the error message provided when an edit can't be translated is not great (like not at all). That's because syntax issues are rare (especially in the big standard setters metafiles) and when I look into providing a better error message, the work was really complicated; it was just not worth it.  That is why when a translation fails (whihc is usually (but not always) because of bad genedits syntax, the process creates a file with the syntax and the first thing I do is to review that syntax.

Anyway.  The error in the standard NAACCR metafile is expected; I saw you reported another issue about that and I will comment in that issue.

I see a lot of errors due to an empty regex. What is the meaning of checking that something follows an "empty" regex? I am not sure what the means or why Genedits allows it...

I also see a lot of errors saying a message was not found; for those, could you maybe open the metafile with the Genedits Writer and verify that those messages indeed exists?

Same for the "unable to find field"; could you open the metafile and confirm you see a field with that exact name?

There are two or three errors that use the "not-so-helpful-translation-error-message", for those, I will probably need the logic of the edit. But let's try to address the easy ones first.
